packages/firestore/test/unit/local/document_overlay_cache.test.ts

```diff
@@ -0,0 +1,257 @@
+    expect(await overlayCache.getOverlay(key('coll/doc1'))).to.equal(null);
+  });
+
+  it('can read saved overlay', async () => {
+    const m = patchMutation('coll/doc1', { 'foo': 'bar' });
+    await saveOverlaysForMutations(2, m);
+    verifyEqualMutations(await overlayCache.getOverlayMutation('coll/doc1'), m);
+  });
+
+  it('can read saved overlays', async () => {
+    const m1 = patchMutation('coll/doc1', { 'foo': 'bar' });
+    const m2 = setMutation('coll/doc2', { 'foo': 'bar' });
+    const m3 = deleteMutation('coll/doc3');
+    await saveOverlaysForMutations(3, m1, m2, m3);
+    verifyEqualMutations(
+      await overlayCache.getOverlayMutation('coll/doc1'),
+      m1
+    );
+    verifyEqualMutations(
+      await overlayCache.getOverlayMutation('coll/doc2'),
+      m2
+    );
+    verifyEqualMutations(
+      await overlayCache.getOverlayMutation('coll/doc3'),
+      m3
+    );
+  });
+
+  it('can overwrite overlays', async () => {
+    const m1 = patchMutation('coll/doc1', { 'foo': 'bar' });
+    const m2 = setMutation('coll/doc1', { 'foo': 'set', 'bar': 42 });
+    await saveOverlaysForMutations(2, m1);
+    await saveOverlaysForMutations(2, m2);
+    verifyEqualMutations(
+      await overlayCache.getOverlayMutation('coll/doc1'),
+      m2
+    );
+  });
+
+  it('can delete overlays repeatedly', async () => {
+    const m = patchMutation('coll/doc1', { 'foo': 'bar' });
+    await saveOverlaysForMutations(2, m);
+    await overlayCache.removeOverlaysForBatchId(2);
+    expect(await overlayCache.getOverlay(key('coll/doc1'))).to.equal(null);
+
+    // Repeat
+    await overlayCache.removeOverlaysForBatchId(2);
+    expect(await overlayCache.getOverlay(key('coll/doc1'))).to.equal(null);
+  });
+
+  it('can get all overlays for collection', async () => {
+    const m1 = patchMutation('coll/doc1', { 'foo': 'bar' });
+    const m2 = setMutation('coll/doc2', { 'foo': 'bar' });
+    const m3 = deleteMutation('coll/doc3');
+    // m4 and m5 are not under "coll"
+    const m4 = setMutation('coll/doc1/sub/sub_doc', { 'foo': 'bar' });
+    const m5 = setMutation('other/doc1', { 'foo': 'bar' });
+    await saveOverlaysForMutations(3, m1, m2, m3, m4, m5);
+
+    const overlays = await overlayCache.getOverlaysForCollection(
+      path('coll'),
+      -1
+    );
+    verifyOverlayContains(overlays, 'coll/doc1', 'coll/doc2', 'coll/doc3');
+  });
+
+  it('can get all overlays since batch ID', async () => {
+    await saveOverlaysForKeys(2, 'coll/doc1', 'coll/doc2');
+    await saveOverlaysForKeys(3, 'coll/doc3');
+    await saveOverlaysForKeys(4, 'coll/doc4');
+
+    const overlays = await overlayCache.getOverlaysForCollection(
+      path('coll'),
+      2
+    );
+    verifyOverlayContains(overlays, 'coll/doc3', 'coll/doc4');
+  });
+
+  it('can get all overlays for collection group', async () => {
+    await saveOverlaysForKeys(2, 'coll1/doc1', 'coll2/doc1');
+    await saveOverlaysForKeys(3, 'coll1/doc2');
```

Comment:
```suggestion
    await saveOverlaysForKeys(3, 'coll2/doc1/coll1/doc2');
```

Slightly better test.

Reply:
👍  YES!